internal/controller/rolebindingtemplate_controller.go

```diff
@@ -0,0 +1,322 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+
+	v1alpha1 "github.com/giantswarm/rbac-operator/api/v1alpha1"
+	pkgkey "github.com/giantswarm/rbac-operator/pkg/key"
+	"github.com/giantswarm/rbac-operator/pkg/project"
+
+	"github.com/giantswarm/k8smetadata/pkg/annotation"
+	"github.com/giantswarm/k8smetadata/pkg/label"
+	security "github.com/giantswarm/organization-operator/api/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// RoleBindingTemplateReconciler reconciles a RoleBindingTemplate object
+type RoleBindingTemplateReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+const (
+	errGetRBT           = "could not get RoleBindingTemplate"
+	errUpdateStatus     = "unable to update status"
+	errNamespacesFailed = "one or more namespaces failed"
+	errGetNamespaces    = "unable to get namespaces from scope"
+	errCreateOrUpdateRB = "could not create or update RoleBinding"
+	errDeleteRB         = "could not delete RoleBinding"
+
+	msgReconcileStarted   = "starting reconciliation"
+	msgReconcileSucceeded = "reconciliation succeeded"
+)
+
+// +kubebuilder:rbac:groups=auth.giantswarm.io,resources=rolebindingtemplates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=auth.giantswarm.io,resources=rolebindingtemplates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=auth.giantswarm.io,resources=rolebindingtemplates/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+// +kubebuilder:rbac:groups="security.giantswarm.io",resources=organizations,verbs=get;list;watch
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=*
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.22.1/pkg/reconcile
+func (r *RoleBindingTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logf.FromContext(ctx)
+
+	// Get the RoleBindingTemplate resource
+	template := &v1alpha1.RoleBindingTemplate{}
+	if err := r.Get(ctx, req.NamespacedName, template); err != nil {
+		if apierrors.IsNotFound(err) {
+			// If the custom resource is not found then it usually means that it was deleted or not created
+			// In this way, we will stop the reconciliation
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, errGetRBT)
+		return ctrl.Result{}, err
+	}
+
+	// Ensure status is updated at the end of reconciliation
+	defer func() {
```

Comment:
Not necessarily objecting, but why is this async? Couldn't this result in the status being updated for partial completion? Couldn't it cause re-reconciliations?

Reply:
This function would also be called if Reconcile() exits due to context cancellation or a panic. I don't _think_ it would cause issues with the current logic, but for example if it were called mid-way through [this loop](https://github.com/giantswarm/rbac-operator/pull/683/changes/BASE..cbe1d60f65a91879d1140a716d845cd3212f2c71#diff-c689a08226834a6ae1337b7f9236772949458feb36279e8b530a83d8c177bf81R180), you'd have incomplete data.